File: gems/server/lib/itsi/server/config/middleware/token_source.rb
module Itsi
  class Server
    module Config

      HeaderSource = TypedStruct.new do
        {
          name: Type(String) & Required(),
          prefix: Type(String)
        }
      end

      HeaderSourceOuter = TypedStruct.new do
        {
          header: Type(HeaderSource)
        }
      end

      QuerySource = TypedStruct.new do
        {
          query: Type(String) & Required()
        }
      end

      TokenSource = TypedStruct.new do
        Or(
          Type(HeaderSourceOuter),
          Type(QuerySource)
        )
      end
    end
  end
end

File: crates/itsi_server/src/server/middleware_stack/middlewares/token_source.rs
use serde::{Deserialize, Serialize};

use crate::server::http_message_types::{HttpRequest, RequestExt};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TokenSource {
    #[serde(rename(deserialize = "header"))]
    Header {
        name: String,
        prefix: Option<String>,
    },
    #[serde(rename(deserialize = "query"))]
    Query(String),
}

impl TokenSource {
    pub fn extract_token<'req>(&self, req: &'req HttpRequest) -> Option<&'req str> {
        match self {
            TokenSource::Header { name, prefix } => req.headers().get(name).and_then(|value| {
                value.to_str().ok().and_then(|value| {
                    if let Some(prefix) = prefix {
                        value.strip_prefix(prefix).map(|v| v.trim())
                    } else {
                        Some(value)
                    }
                })
            }),
            TokenSource::Query(query_name) => req.query_param(query_name),
        }
    }
}
